File: Projects/ShopifyLarkTracker/ShopifyLarkTracker/Services/ShopifyService.cs
using Microsoft.Extensions.Configuration;
using ShopifyLarkTracker.Models;
using System.Diagnostics;
using System.Net.Http.Headers;
using System.Text.Json;

namespace ShopifyLarkTracker.Services;

public class ShopifyService
{
    private readonly IConfiguration _config;
    private readonly HttpClient _http;

    public ShopifyService(IConfiguration config)
    {
        _config = config;
        _http = new HttpClient();
    }

    public async Task<List<ShopifyOrder>> GetOrdersAsync(
        DateTime? updatedAfter = null)
    {
        var store =
            _config["Shopify:StoreUrl"];

        var token =
            _config["Shopify:AccessToken"];

        if (string.IsNullOrWhiteSpace(store))
            throw new Exception(
                "Shopify StoreUrl missing");

        if (string.IsNullOrWhiteSpace(token))
            throw new Exception(
                "Shopify AccessToken missing");

        var url =
            $"{store}/admin/api/2025-01/orders.json?status=any&limit=250";

        if (updatedAfter.HasValue)
        {
            url +=
                $"&updated_at_min={Uri.EscapeDataString(
                    updatedAfter.Value
                        .ToUniversalTime()
                        .ToString("O"))}";
        }

        Console.WriteLine(
            $"SHOPIFY URL = {url}");

        var request =
            new HttpRequestMessage(
                HttpMethod.Get,
                url);

        request.Headers.Add(
            "X-Shopify-Access-Token",
            token);

        request.Headers.Accept.Add(
            new MediaTypeWithQualityHeaderValue(
                "application/json"));

        var apiTimer =
            Stopwatch.StartNew();

        var response =
            await _http.SendAsync(request);

        apiTimer.Stop();

        Console.WriteLine(
            $"SHOPIFY API TIME = {apiTimer.Elapsed.TotalSeconds:F2}s");

        Console.WriteLine(
            $"SHOPIFY STATUS CODE = {(int)response.StatusCode}");

        var json =
            await response.Content.ReadAsStringAsync();

        Console.WriteLine(
            $"SHOPIFY RESPONSE SIZE = {json.Length:N0} chars");

        response.EnsureSuccessStatusCode();

        var parseTimer =
            Stopwatch.StartNew();

        using var doc =
            JsonDocument.Parse(json);

        var result =
            new List<ShopifyOrder>();

        if (!doc.RootElement.TryGetProperty(
                "orders",
                out var orders))
        {
            return result;
        }

        foreach (var order in
            orders.EnumerateArray())
        {
            result.Add(
                new ShopifyOrder
                {
                    ShopifyId =
                        long.TryParse(
                            GetString(order, "id"),
                            out var shopifyId)
                                ? shopifyId
                                : 0,

                    OrderNumber =
                        GetString(
                            order,
                            "order_number"),

                    SourceName =
                        GetString(
                            order,
                            "source_name"),

                    Status =
                        GetFulfillmentStatus(
                            order),

                    CreatedAt =
                        GetDate(
                            order,
                            "created_at"),

                    Customer =
                        ExtractCustomer(
                            order),

                    ShippingAddress =
                        ExtractShippingAddress(
                            order),

                    Fulfillments =
                        ExtractFulfillments(
                            order)
                });
        }

        parseTimer.Stop();

        Console.WriteLine(
            $"Loaded {result.Count} Shopify orders");

        Console.WriteLine(
            $"SHOPIFY PARSE TIME = {parseTimer.Elapsed.TotalSeconds:F2}s");

        return result;
    }

    private ShopifyCustomer? ExtractCustomer(
        JsonElement order)
    {
        if (!order.TryGetProperty(
                "customer",
                out var customer))
        {
            return null;
        }

        if (customer.ValueKind ==
            JsonValueKind.Null)
        {
            return null;
        }

        return new ShopifyCustomer
        {
            FirstName =
                GetString(
                    customer,
                    "first_name"),

            LastName =
                GetString(
                    customer,
                    "last_name")
        };
    }

    private ShopifyShippingAddress?
        ExtractShippingAddress(
        JsonElement order)
    {
        if (!order.TryGetProperty(
                "shipping_address",
                out var address))
        {
            return null;
        }

        if (address.ValueKind ==
            JsonValueKind.Null)
        {
            return null;
        }

        return new ShopifyShippingAddress
        {
            Name =
                GetString(
                    address,
                    "name")
        };
    }

    private List<ShopifyFulfillment>
        ExtractFulfillments(
        JsonElement order)
    {
        var result =
            new List<ShopifyFulfillment>();

        if (!order.TryGetProperty(
                "fulfillments",
                out var fulfillments))
        {
            return result;
        }

        foreach (var f in
            fulfillments.EnumerateArray())
        {
            result.Add(
                new ShopifyFulfillment
                {
                    TrackingNumber =
                        GetString(
                            f,
                            "tracking_number"),

                    CreatedAt =
                        GetDate(
                            f,
                            "created_at")
                });
        }

        return result;
    }

    private string GetFulfillmentStatus(
        JsonElement order)
    {
        if (order.TryGetProperty(
                "fulfillment_status",
                out var status))
        {
            return status.ToString();
        }

        return "unknown";
    }

    private DateTime GetDate(
        JsonElement element,
        string property)
    {
        if (!element.TryGetProperty(
                property,
                out var value))
        {
            return DateTime.UtcNow;
        }

        if (DateTime.TryParse(
                value.ToString(),
                out var date))
        {
            return date;
        }

        return DateTime.UtcNow;
    }

    private string GetString(
        JsonElement element,
        string property)
    {
        if (element.ValueKind !=
            JsonValueKind.Object)
        {
            return "";
        }

        if (!element.TryGetProperty(
                property,
                out var value))
        {
            return "";
        }

        switch (value.ValueKind)
        {
            case JsonValueKind.String:
                return value.GetString() ?? "";

            case JsonValueKind.Number:
                return value.ToString();

            case JsonValueKind.True:
                return "true";

            case JsonValueKind.False:
                return "false";

            case JsonValueKind.Null:
                return "";

            default:
                return value.ToString();
        }
    }
}

File: Projects/ShopifyLarkTracker/logivis-worker/src/services/shopify.service.ts
import {
	ShopifyOrder,
	ShopifyCustomer,
	ShopifyShippingAddress,
	ShopifyFulfillment,
	ShopifyLineItem
} from "../models/shopify";



interface ShopifyOrdersResponse {
	orders: any[];
}
export class ShopifyService {

	constructor(
		private readonly env: Env
	) {
	}

	async getOrders(
		updatedAfter?: Date
	): Promise<ShopifyOrder[]> {

		console.log("Entered getOrders()");

		const store = this.env.SHOPIFY_STORE;
		const token = this.env.SHOPIFY_ACCESS_TOKEN;

		if (!store)
			throw new Error("SHOPIFY_STORE missing");

		if (!token)
			throw new Error("SHOPIFY_ACCESS_TOKEN missing");

		let url =
			`${store}/admin/api/2025-01/orders.json?status=any&limit=250`;

		if (updatedAfter) {

			url +=
				`&updated_at_min=${encodeURIComponent(
					updatedAfter.toISOString()
				)}`;
		}

		console.log(`SHOPIFY URL = ${url}`);

		const apiStart = Date.now();

		const response =
			await fetch(url, {
				method: "GET",
				headers: {
					"X-Shopify-Access-Token": token,
					"Accept": "application/json"
				}
			});

		console.log(
			`SHOPIFY API TIME = ${((Date.now() - apiStart) / 1000).toFixed(2)}s`
		);

		console.log(
			`SHOPIFY STATUS = ${response.status}`
		);

		if (!response.ok) {

			throw new Error(
				await response.text()
			);
		}

		const json =
			await response.text();

		console.log(
			`SHOPIFY RESPONSE SIZE = ${json.length.toLocaleString()} chars`
		);

		const parseStart =
			Date.now();

		const root =
			JSON.parse(json);

		const result: ShopifyOrder[] = [];

		const orders =
			root.orders ?? [];

		for (const order of orders) {

			result.push({

				shopifyId:
					Number(order.id) || 0,

				orderNumber:
					String(order.order_number ?? ""),

				sourceName:
					String(order.source_name ?? ""),

				status:
					order.fulfillment_status ??
					"unknown",

				createdAt:
					order.created_at
						? new Date(order.created_at)
						: new Date(),

				customer:
					this.extractCustomer(order),

				shippingAddress:
					this.extractShippingAddress(order),

				lineItems:
					this.extractLineItems(order),

				fulfillments:
					this.extractFulfillments(order)

			});
		}

		console.log(
			`Loaded ${result.length} Shopify orders`
		);

		console.log(`Loaded ${result.length} Shopify orders`);

		const order15324 = result.find(
			x => x.orderNumber === "15324"
		);

		console.log(
			`Contains Order 15324: ${order15324 != null}`
		);

		if (order15324) {

			console.log("================================");
			console.log("Order 15324 Found");
			console.log("================================");

			console.log(JSON.stringify(order15324, null, 2));

		}

		console.log(
			`SHOPIFY PARSE TIME = ${((Date.now() - parseStart) / 1000).toFixed(2)}s`
		);

		return result;
	}

	async getOrderByOrderNo(
		orderIds: string
	): Promise<ShopifyOrder | null> {

		console.log("Entered getOrderByOrderNo()");

		const store = this.env.SHOPIFY_STORE;
		const token = this.env.SHOPIFY_ACCESS_TOKEN;

		if (!store)
			throw new Error("SHOPIFY_STORE missing");

		if (!token)
			throw new Error("SHOPIFY_ACCESS_TOKEN missing");

		const normalized =
			this.normalizeShopifyOrderNo(orderIds);

		const url =
			`${store}/admin/api/2025-01/orders.json?status=any&name=${encodeURIComponent(normalized)}&limit=1`;

		console.log(`SHOPIFY URL = ${url}`);

		const response =
			await fetch(url, {
				method: "GET",
				headers: {
					"X-Shopify-Access-Token": token,
					"Accept": "application/json"
				}
			});

		console.log(
			`SHOPIFY STATUS = ${response.status}`
		);

		if (!response.ok) {

			throw new Error(
				await response.text()
			);

		}

		const root = await response.json() as ShopifyOrdersResponse;

		const order =
			root.orders?.[0];

		if (!order) {

			console.log(`Order ${normalized} not found.`);

			return null;

		}

		return {

			shopifyId:
				Number(order.id) || 0,

			orderNumber:
				String(order.order_number ?? ""),

			sourceName:
				String(order.source_name ?? ""),

			status:
				order.fulfillment_status ??
				"unknown",

			createdAt:
				order.created_at
					? new Date(order.created_at)
					: new Date(),

			customer:
				this.extractCustomer(order),

			shippingAddress:
				this.extractShippingAddress(order),

			lineItems:
				this.extractLineItems(order),

			fulfillments:
				this.extractFulfillments(order)

		};

	}

	private normalizeShopifyOrderNo(
		orderNo: string
	): string {

		const value = orderNo.trim();

		if (!value) {
			return "";
		}

		let normalized =
			value.replace(/^#/, "");

		normalized =
			normalized.replace(/-\d+$/, "");

		return `#${normalized}`;

	}

	private extractCustomer(
		order: any
	): ShopifyCustomer | null {

		const customer =
			order.customer;

		if (!customer)
			return null;

		return {

			firstName:
				this.getString(customer, "first_name"),

			lastName:
				this.getString(customer, "last_name")

		};
	}

	private extractShippingAddress(
		order: any
	): ShopifyShippingAddress | null {

		const address =
			order.shipping_address;

		if (!address)
			return null;

		return {

			name:
				this.getString(address, "name"),

			phone:
				this.getString(address, "phone"),

			address1:
				this.getString(address, "address1"),

			address2:
				this.getString(address, "address2"),

			city:
				this.getString(address, "city"),

			province:
				this.getString(address, "province"),

			zip:
				this.getString(address, "zip"),

			country:
				this.getString(address, "country")

		};
	}

	private extractFulfillments(
		order: any
	): ShopifyFulfillment[] {

		const result: ShopifyFulfillment[] = [];

		const fulfillments =
			order.fulfillments ?? [];

		for (const f of fulfillments) {

			result.push({

				trackingNumber:
					this.getString(f, "tracking_number"),

				createdAt:
					this.getDate(f, "created_at")

			});
		}

		return result;
	}

	private extractLineItems(
		order: any
	): ShopifyLineItem[] {

		const result: ShopifyLineItem[] = [];

		const items = order.line_items ?? [];

		for (const item of items) {

			result.push({

				id:
					Number(item.id ?? 0),

				title:
					this.getString(item, "title"),

				sku:
					this.getString(item, "sku"),

				quantity:
					Number(item.quantity ?? 0),

				variantTitle:
					this.getString(item, "variant_title"),

				grams:
					Number(item.grams ?? 0),

				requiresShipping:
					Boolean(item.requires_shipping)

			});

		}

		return result;

	}

	private getDate(
		obj: any,
		property: string
	): Date {

		const value =
			obj?.[property];

		if (!value)
			return new Date();

		const date =
			new Date(value);

		if (isNaN(date.getTime()))
			return new Date();

		return date;
	}

	private getString(
		obj: any,
		property: string
	): string {

		const value =
			obj?.[property];

		if (
			value === undefined ||
			value === null
		) {
			return "";
		}

		return String(value);
	}

}
